Enable user-defined aggregates

psarna opened the pull request:
It turns out that user-defined aggregates did not need any elaborate coding in order to make them exposed to the users. The whole infrastructure is already there, including system schema tables and support for running aggregate queries, so this series simply adds lots and lots of boilerplate glue code to make UDA usable.

It also comes with a simple test which shows that it's possible to define and use such an aggregate.

Performance not tested, since user-defined functions are still experimental, so nothing really changes in this matter.

Tests: unit(release)

Fixes #7201

nyh commented:
Please Fixes or at least mention https://github.com/scylladb/scylla/issues/7201

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +494,32 @@ static shared_ptr<aggregate_function> make_count_function() {
 
+// Drops the first arg type from the types declaration (which denotes the accumulator)
+// in order to compute the actual type of given user-defined-aggregate (UDA)
+static std::vector<data_type> state_arg_types_to_uda_arg_types(const std::vector<data_type> &arg_types) {
```
`&` stuck to the arg instead of the types. Maybe it was confused by the arg named "arg_types".

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +494,32 @@ static shared_ptr<aggregate_function> make_count_function() {
+// Drops the first arg type from the types declaration (which denotes the accumulator)
+// in order to compute the actual type of given user-defined-aggregate (UDA)
+static std::vector<data_type> state_arg_types_to_uda_arg_types(const std::vector<data_type> &arg_types) {
+    assert(arg_types.size() >= 2);
```
on_internal_error()

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +494,32 @@ static shared_ptr<aggregate_function> make_count_function() {
+static std::vector<data_type> state_arg_types_to_uda_arg_types(const std::vector<data_type> &arg_types) {
+    assert(arg_types.size() >= 2);
+    std::vector<data_type> types;
+    types.reserve(arg_types.size() - 1);
```
(I don't think it's really needed, vector probably does this itself)

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +494,32 @@ static shared_ptr<aggregate_function> make_count_function() {
+        : abstract_function(std::move(fname), state_arg_types_to_uda_arg_types(sfunc->arg_types()), finalfunc->return_type())
+        , _initcond(initcond)
+        , _sfunc(sfunc)
+        , _finalfunc(finalfunc)
```
std::move()s?

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +494,32 @@ static shared_ptr<aggregate_function> make_count_function() {
+    return std::make_unique<impl_user_aggregate>(_initcond, _sfunc, _finalfunc);
+}
+
+bool user_aggregate::is_pure() const { return true; }
```
Better to refer to _sfunc and _finalfunc in case one of them is not pure.

avikivity commented on cql3/functions/user_aggregate.hh:
```diff
@@ -0,0 +1,57 @@
+ * Copyright (C) 2021-present ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
```
No need for these Modified blurbs.

avikivity commented on cql3/selection/abstract_function_selector.hh:
```diff
@@ -66,7 +67,8 @@ public:
             : _fun(std::move(fun)), _arg_selectors(std::move(arg_selectors)),
-              _requires_thread(boost::algorithm::any_of(_arg_selectors, [] (auto& s) { return s->requires_thread(); })) {
+              _requires_thread(boost::algorithm::any_of(_arg_selectors, [] (auto& s) { return s->requires_thread(); })
+                    || bool(dynamic_pointer_cast<cql3::functions::user_aggregate>(_fun))) {
```
Better to ask the function if it requires a thread than to assume it. I see abstract_function has a requires_thread() function.

psarna replied:
Oh, right, it actually looks like a preexisting bug... the function itself should be asked anyway.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1868,6 +1902,45 @@ std::vector<mutation> make_drop_function_mutations(shared_ptr<cql3::functions::u
+}
+
+std::vector<mutation> make_drop_aggregate_mutations(shared_ptr<cql3::functions::user_aggregate> aggregate, api::timestamp_type timestamp) {
+    return make_drop_function_mutations(aggregates(), *aggregate, timestamp);
```
I think we need to check that we don't drop a scalar function if it's used in an aggregate.

psarna replied:
Indeed, I also need to add lots of test cases, including ones that drop and recreate state and final functions.

avikivity commented on cql3/statements/create_aggregate_statement.cc:
```diff
@@ -0,0 +1,86 @@
+    if (!_aggregate) {
+        return make_ready_future<::shared_ptr<cql_transport::event::schema_change>>();
+    }
+    return qp.get_migration_manager().announce_new_aggregate(_aggregate).then([this] {
```
Should we block this on a schema feature? It's painful, but on the other hand someone creating an aggregate before the cluster is fully upgraded can easily confuse it beyond all repair.

psarna replied:
Probably, we didn't create any feature for UDF-s though, and it's a nearly identical case. Perhaps we ignored the schema feature problem because the support is still marked as experimental? Or maybe not doing so was a bug.

psarna replied:
Oh wait, we didn't create a **schema** feature for UDF-s, but there is a cluster feature. So I'll just add one for UDA as well.

avikivity commented on test/cql-pytest/test_uda.py:
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020-present ScyllaDB
```
2021

avikivity commented on test/cql-pytest/test_uda.py:
```diff
@@ -0,0 +1,39 @@
+            with new_aggregate(cql, test_keyspace, custom_avg_body) as custom_avg:
+                custom_res = [row for row in cql.execute(f"SELECT {test_keyspace}.{custom_avg}(id) AS result FROM {table}")]
+                avg_res = [row for row in cql.execute(f"SELECT avg(id) AS result FROM {table}")]
+                assert custom_res == avg_res
```
Please add tests that you can't create an aggregate if scalar components are missing, can't drop scalars if they're used, and that you can't use the aggregate after it's dropped. Also mismatches of argument types etc.

psarna commented:
(I addressed part of the comments, but the most important one - more tests, remains to be done - so I'm not marking it a "v2" yet)

psarna commented:
v2:
 - added a UDA cluster feature which prevents creating user-defined aggregates on non-upgraded clusters
 - ensured that functions used by aggregates can't be dropped before the aggregate itself is dropped
 - added more tests

nyh commented:
Looks good, I think you addressed all of Avi's comments and I can commit now, but I'll let you see first if you want to do anything about my small nitpicks.

psarna commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -468,6 +498,33 @@ static shared_ptr<aggregate_function> make_count_function() {
+// in order to compute the actual type of given user-defined-aggregate (UDA)
+static std::vector<data_type> state_arg_types_to_uda_arg_types(const std::vector<data_type>& arg_types) {
+    if(arg_types.size() < 2) {
+        on_internal_error(cql3::functions::log, "State function for user-defined aggregates needs at least two arguments");
```
hm, let me check what happens if I define the state function as having more arguments -- perhaps it's a good candidate for a new test case.

psarna replied:
edit: it is an internal error. When creating the aggregate via CQL, a state function with very specific signature is looked up - with a parameter list synthesized from the state type and the arg type specified by the aggregate. It is therefore not possible to end up calling `state_arg_types_to_uda_arg_types` on a function which takes a different number of parameters than 2, unless it's a programmer's mistake. Actually, the internal error should be even stricter and throw if the number of args `!= 2`, not just `< 2`.

psarna commented on cql3/functions/functions.cc:
```diff
@@ -158,6 +160,16 @@ void functions::remove_function(const function_name& name, const std::vector<dat
     with_udf_iter(name, arg_types, [] (functions::declared_t::iterator i) { _declared.erase(i); });
 }
 
+std::optional<function_name> functions::used_by_user_aggregate(const function_name& name) {
```
Yup, the usage was meant to be in this patch, it probably got mixed up during a rebase. Thanks

psarna commented:
v3:
 - one more test (for a state function which was declared with an incorrect number of arguments)
 - removed redundant `#if 0` blocks
 - hardened the internal error condition - the state transition function now expects exactly 2 arguments, not "at least 2"

Rhyanz46 commented:
wow, it's done?  I just a junior developer,, can't read code perfectly

psarna commented:
> wow, it's done? I just a junior developer,, can't read code perfectly

It is, but note that a fundamental building block for user-defined aggregates are user-defined functions, which are still experimental and require two steps to be usable:
 - turning on an experimental feature: `experimental_features:udf`
 - enabling them explicitly in yaml: `enable_user_defined_functions`

Also, user-defined aggregates are not part of any official release yet, and they're still in the experimental phase themselves.
